### PR TITLE
do a single pass in maxByFirstDiffering

### PR DESCRIPTION
**Pull request by josharian**
This cuts the runtime for hat allocation
in my tests by about 20%.

No functional changes.

I have not added any tests, but I have confirmed
that there are no functional changes
using my not-yet-submitted hat golden tests.

Part of the reason that the hat golden tests are not
yet ready for review is that they are way too slow.

## Checklist

- [/] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [/] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [/] I have not broken the cheatsheet

### CODE REVIEW

**Comment by josharian** [comment]
marking as draft because I think there may be a deeper fix...

**Comment by josharian** [comment]
this appears to be cumulative with my other fixes, ready for review

File: packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.ts [hunk 1]
```diff
@@ -30,8 +30,22 @@ export function maxByFirstDiffering<T>(
...
+      if (value < best) {
+        continue;
+      }
+      if (value > best) {
+        best = value;
+        keep.length = 0;
+      }
+      keep.push(item);
+    }
+    remainingValues = keep;
```
Comment (pokey) [inline packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.ts:48]: I would be tempted to make this its own function for clarity, but I don't feel strongly. I guess there may be performance implications if this code is on the hot path

**Review by pokey — APPROVED** [review]
Looks good with one minor comment!

File: packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.ts (hunk 1 above)
Comment (josharian) [reply packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.ts]: Done. I'm still finding my feet for typescript performance, but I'm pretty confident that the JIT will figure out how to inline this. And empirically it is just as fast.

**Comment by josharian** [comment]
What is the convention for PRs that are approved with comments? wait for another round of review to confirm you like the fixes?

File: packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.ts [hunk 2]
```diff
@@ -24,15 +24,45 @@ export function maxByFirstDiffering<T>(
...
 }
+
+/**
+ * Given an array of items and a function that returns a number for each item,
+ * return all items that share the maximum value according to that function.
+ * @param arr The array to find the max values of
+ * @param fn A function that returns a number for each item in the array
+ * @returns All items in the array that share the maximum value
+ **/
+function highest<T>(arr: T[], fn: (item: T) => number): T[] {
```
Comment (pokey) [inline packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.ts:43]: Maybe `maxByAllowingTies` instead of `highets`?

**Review by pokey — APPROVED** [review]
Minor quibble; otherwise good.  I would be very tempted to add unit tests for this function, but I'm hesitant to make you invest more in this PR, which was supposed to be a quick win

If you think not worth it to add tests, and you're happy with my rename in comment below, feel free to merge

**Comment by pokey** [comment]
> What is the convention for PRs that are approved with comments? wait for another round of review to confirm you like the fixes?

It's a good question. I think if the comments resulted in changes that don't really have any room for disagreement, then feel free to merge. In this case, I think there's room for disagreement over the function name (see my comment 😄), so maybe better to leave it?

**Comment by josharian** [comment]
> I think if the comments resulted in changes that don't really have any room for disagreement, then feel free to merge.

OK. Although given the naturally widely divergent range of opinions, I would have a weak preference for only approving once you are sold that the only remaining changes are trivial.

Will rename and will probably add tests. Thanks for the speedy review.

**Comment by pokey** [comment]
> OK. Although given the naturally widely divergent range of opinions, I would have a weak preference for only approving once you are sold that the only remaining changes are trivial.

yeah fwiw that's what I usually do. in this case I approved because I wasn't sure whether it was worth making the changes at all

**Comment by josharian** [comment]
since I'm new to the project and to typescript, I'm going to be more annoying about asking for re-reviews rather than just merging. so: PTAL. :)

File: packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.test.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      [0, -Infinity],
+      [0, 1, 0, 1],
+      [1, 0, 1, 0],
+      [0, 1, 1, 0],
+      [1, 0, 0, 1],
+    ];
+
+    testCases.forEach((testCase) => {
+      const actual = maxByAllowingTies(testCase, (x) => x);
+      const expected = goldenMaxByAllowingTies(testCase, (x) => x);
```
Comment (pokey) [inline packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.test.ts:36]: Cute. I like it

**Review by pokey — APPROVED** [review]
Looks good!

File: packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.test.ts (hunk 3 above)
Comment (josharian) [reply packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.test.ts:36]: An old favorite approach from fuzz testing. Particularly well suited to performance optimization. :)

https://en.wikipedia.org/wiki/Differential_testing

**Pull request merged**